botorch/optim/fit.py

```diff
@@ -179,119 +183,80 @@ def store_iteration(xk):
+    if closure is None:
+        closure = get_loss_closure(mll)
 
-    # update with user-supplied bounds (overwrites if already exists)
+    param_dict, bounds_dict = get_parameters_and_bounds(mll, requires_grad=True)
     if bounds is not None:
-        bounds_.update(bounds)
+        bounds_dict.update(bounds)
 
-    iterations = []
-    t1 = monotonic()
+    if stopping_criterion is DEFAULT:
+        stopping_criterion = ExpMAStoppingCriterion()
 
-    param_trajectory: Dict[str, List[Tensor]] = {
-        name: [] for name, param in mll.named_parameters()
-    }
-    loss_trajectory: List[float] = []
-    i = 0
-    stop = False
-    stopping_criterion = ExpMAStoppingCriterion(
-        **_filter_kwargs(ExpMAStoppingCriterion, **optim_options)
-    )
-    train_inputs, train_targets = mll.model.train_inputs, mll.model.train_targets
-    while not stop:
-        optimizer.zero_grad()
+    if isinstance(torch_optimizer, (type, partial)):
+        torch_optimizer = torch_optimizer(params=list(param_dict.values()))
+
+    loss = None
+    step_counter = count()
+    for step in step_counter:
+        if step_limit is not None and step >= step_limit:
+            break
+
+        torch_optimizer.zero_grad()
         with gpt_settings.fast_computations(log_prob=approx_mll):
-            output = mll.model(*train_inputs)
-            # we sum here to support batch mode
-            args = [output, train_targets] + _get_extra_mll_args(mll)
-            loss = -mll(*args).sum()
-            loss.backward()
-        loss_trajectory.append(loss.item())
-        for name, param in mll.named_parameters():
-            param_trajectory[name].append(param.detach().clone())
-        if optim_options["disp"] and (
-            (i + 1) % 10 == 0 or i == (optim_options["maxiter"] - 1)
-        ):
-            print(f"Iter {i + 1}/{optim_options['maxiter']}: {loss.item()}")
-        if track_iterations:
-            iterations.append(OptimizationIteration(i, loss.item(), monotonic() - t1))
-
-        optimizer.step()
-        # project onto bounds:
-        if bounds_:
-            for pname, param in mll.named_parameters():
-                if pname in bounds_:
-                    param.data = param.data.clamp(*bounds_[pname])
-        i += 1
-        stop = stopping_criterion.evaluate(fvals=loss.detach())
-    info_dict = {
-        "fopt": loss_trajectory[-1],
-        "wall_time": monotonic() - t1,
-        "iterations": iterations,
-    }
-    return mll, info_dict
+            loss = closure()
+
+        if callback:
+            callback(OptimizationIteration(step, loss, monotonic() - start_time))
+
+        if stopping_criterion and stopping_criterion.evaluate(fvals=loss.detach()):
+            break
+
+        loss.backward()
+        torch_optimizer.step()
+        for name, limits in bounds_dict.items():
+            param = param_dict[name]
+            param.data = param.clamp(*limits)
+
+        if torch_scheduler is not None:
```

Comment:
The docstring suggests `torch_scheduler` could also be a factory, but I don't see where that's checked, and here it's assumed to already have `step`